drivers/flash/flash_mcux_flexspi_nor.c

```diff
@@ -494,11 +459,583 @@ static void flash_flexspi_nor_pages_layout(const struct device *dev,
+{
+	uint32_t flexspi_lut[FLEXSPI_INSTR_END][MEMC_FLEXSPI_CMD_PER_SEQ] = {0};
+	/* JESD216B defines up to 23 basic flash parameters */
+	uint32_t param_buf[23];
+	/* Space to store SFDP header and first parameter header */
+	uint8_t sfdp_buf[JESD216_SFDP_SIZE(1)] __aligned(4);
+	struct jesd216_bfp *bfp = (struct jesd216_bfp *)param_buf;
+	struct jesd216_sfdp_header *header = (struct jesd216_sfdp_header *)sfdp_buf;
+	int ret;
+	unsigned int key = 0U;
+
+	flexspi_device_config_t config = {
+		.flexspiRootClk = MHZ(50),
+		.flashSize = FLEXSPI_FLSHCR0_FLSHSZ_MASK, /* Max flash size */
+		.ARDSeqNumber = 1,
+		.ARDSeqIndex = READ,
+	};
+
+	if (memc_flexspi_is_running_xip(&data->controller)) {
+		/*
+		 * ==== ENTER CRITICAL SECTION ====
+		 * No flash access should be performed in critical section. All
+		 * code and data accessed must reside in ram.
+		 */
+		key = irq_lock();
+		memc_flexspi_wait_bus_idle(&data->controller);
+	}
+
+	/* SFDP spec requires that we downclock the FlexSPI to 50MHz or less */
+	ret = memc_flexspi_update_clock(&data->controller, &config,
+					data->port, MHZ(50));
+	if (ret < 0) {
+		goto _exit;
+	}
+
+	/* Setup initial LUT table and FlexSPI configuration */
+	memcpy(flexspi_lut, flash_flexspi_nor_base_lut, sizeof(flash_flexspi_nor_base_lut));
+
+	ret = memc_flexspi_set_device_config(&data->controller, &config,
+					(uint32_t *)flexspi_lut,
+					FLEXSPI_INSTR_END * MEMC_FLEXSPI_CMD_PER_SEQ,
+					data->port);
+	if (ret < 0) {
+		goto _exit;
+	}
+
+	/* First, check if the JEDEC ID of this flash has explicit support
+	 * in this driver
+	 */
+	ret = flash_flexspi_nor_check_jedec(data, flexspi_lut);
+	if (ret == 0) {
+		/* Flash was supported, SFDP probe not needed */
+		goto _program_lut;
+	}
+
+	ret = flash_flexspi_nor_sfdp_read_helper(data, 0, sfdp_buf, sizeof(sfdp_buf));
+	if (ret < 0) {
+		goto _exit;
+	}
+
+	LOG_DBG("SFDP header magic: 0x%x", header->magic);
+	if (jesd216_sfdp_magic(header) != JESD216_SFDP_MAGIC) {
+		/* Header was read incorrectly */
+		LOG_WRN("Invalid header, using legacy SPI mode");
+		data->legacy_poll = true;
+		goto _program_lut;
+	}
+
+	if (header->phdr[0].len_dw > ARRAY_SIZE(param_buf)) {
+		/* Not enough space to read parameter table */
+		ret = -ENOBUFS;
+		goto _exit;
+	}
+
+	/* Read basic flash parameter table */
+	ret = flash_flexspi_nor_sfdp_read_helper(data,
+			jesd216_param_addr(&header->phdr[0]),
+			param_buf,
+			sizeof(uint32_t) * header->phdr[0].len_dw);
+	if (ret < 0) {
+		goto _exit;
+	}
+
+	/* Configure flash */
+	ret = flash_flexspi_nor_config_flash(data, header, bfp, flexspi_lut);
+	if (ret < 0) {
+		goto _exit;
+	}
+
+_program_lut:
+	/*
+	 * Update the FlexSPI with the config structure provided
+	 * from devicetree and the configured LUT
+	 */
+	ret = memc_flexspi_set_device_config(&data->controller, &data->config,
+					(uint32_t *)flexspi_lut,
+					FLEXSPI_INSTR_END * MEMC_FLEXSPI_CMD_PER_SEQ,
+					data->port);
+	if (ret < 0) {
+		return ret;
+	}
```

Comment:
At this point you leave the `irq_lock()` from line 945 in effect, is that correct?

Reply:
Yes, this is desired- if the flash configuration block cannot be programmed to the FlexSPI at this point (and we are running in execute in place mode) further execution of the init flow is likely impossible, as the FlexSPI won't be able to access flash. Therefore we return with the IRQ lock held, and drop into the `while 1` loop on line 1057. This is preferable to continuing execution, because it gives a developer a chance to debug the failure, while continuing the init flow would likely crash the chip.